lib/new_relic/agent/instrumentation/redis/instrumentation.rb

```diff
@@ -94,5 +87,17 @@ def _nr_redis_client_config
         config
       end
     end
+
+    def _nr_db
+      # db is a method on the Redis client in versions < 5.x
+      return db if respond_to?(:db)
+
+      begin
+        _nr_redis_client_config.db
+      rescue StandardError => e
+        NewRelic::Agent.logger.error("Failed to determine configured Redis db value: #{e.class} - #{e.message}")
```

Comment:
is db something that could be memoized? Like would it be the same all the time for a specific instance, or could the db be different within the same instance? I'm mainly thinking about this in the context of other instrumentation we've done recently (like the dynamodb one), where customers were seeing log messages flood their logs when the client couldnt get the something.

Also, does it make sense that this would be an error level log, or would it be more of a warn or debug situation?

Reply:
Good question! 

It looks like versions below 5.x could have the DB change at any time [(there's a setter and a getter)](https://github.com/redis/redis-rb/blob/v4.8.1/lib/redis/client.rb#L73-L79), so we shouldn't memoize those. 

However, versions 5.x+ don't have a setter for the DB and don't seem to have any expectations where a DB would change for a given config object [(here are the tests)](https://github.com/redis-rb/redis-client/blob/master/test/redis_client/config_test.rb). I think we should expect it to be the same for the given config/client.

Since we're already memoizing the `_nr_redis_client_config`, I'm not sure we'd get extra performance from memoizing the `db` value within that object. 

Despite this, we're still in a situation where lots of errors could be logged. 

I think you're right about the log level. It looks like similar messages are at both the `warn` and `debug` levels. I think we could put it in either, but maybe making it `debug` would limit our chances of causing a log flooding problem. Do you have a preference?

We could also change the code to run `_nr_redis_client_config.db` to only execute if the object responds to `db`. Do you prefer that approach?

Reply:
That makes sense, thank you!

As for log level, i think based on issues filed relating to a lot of log msgs slowing customers down, I think debug would probably be the best bet, just in case.

I do also like your suggestion of just adding a responds_to instead though, if that seems like a good idea to you. db is allowed to be nil on datastore segments, so it not being available is still ok.

I think either a debug log level or responds_to are both good options to avoid the situation i was thinking about though, so whichever one you prefer is cool with me!

Reply:
Nice! This all sounds good to me. Updated in 337f88842